### ISSUE TITLE
Indoor Spotlight--Subway Station

### ISSUE THREAD

**Issue by Grubleafeater** [CleverRaven/Cataclysm-DDA#42454]
# Describe the bug

A spotlight appears indoors at a subway station.

# Expected behavior
Spotlights should not be able to appear indoors

# Screenshots
https://cdn.discordapp.com/attachments/598523535169945607/736733462286958602/unknown.png
# Additional context
Someone mentioned skylights. Spotlight might be under a skylight.

**Comment by Duck13Tales** [comment]
I have a milspec that lights inside shops, houses etc too. If you wish I have the save.

**Comment by MarlossCDDA** [comment]
I don't think skylights are the issue here - just started a game where there's a roadblock next to the shelter and the searchlight is visible inside the shelter with all windows closed, and I haven't opened the door to the roof.

**Comment by Maleclypse** [comment]
Knowing very little about the vision code, I would imagine that the spotlight probably needs the same early exit that monster aggression uses to ignore things it can't see.  But that's all deep magic to me outside of the concepts.

**Comment by LyleSY** [comment]
Looked at the mapgen JSON and this isn't thrown by that. This is coming from outside and being projected in, I'm guessing from the SEARCHLIGHT monster special attack from a milspec searchlight. That doesn't appear in the monster_special_attacks JSON so I'm guessing that it's hardcoded. Based on what I know, this is behaving as intended but I'm happy to be educated on this.

**Comment by kevingranade** [comment]
They mean the light from the spotlight shouldn't be able to shine into the subway.

**Comment by captainego** [comment]
After reading through this code for a bit, and running a couple of tests, the searchlight checks if it can "see" the character correctly, but it does not check if it can "see" the location where the light is going to point next (the potential target location).  

A side effect of not checking if the searchlight can see the potential target location is that the searchlight places a "field" fd_spotlight centered at the potential target location.  The field does not appear to be "projected" from the searchlight (but the shape is influenced by the location of the searchlight), so the field effect (in this case light) appears even though light should not be able to project from the searchlight to the given point.

I have played with a couple of solutions where I prevent the searchlight from moving to a potential location that it cannot "see".  These solutions solve the problem of the searchlight light illuminating `incorrect locations, but I have seen some interesting behaviors in scenarios where there are a lot of visual obstacles for the searchlight.

I am going to investigate the "field" code for a bit to see if I can see a better solution in that area.  If not I will refine the changes I have made to mattack::searchlight and submit a PR.
